modules/core/src/webgl/classes/program.js

```diff
@@ -214,6 +195,136 @@ export default class Program extends Resource {
   // stub for shader chache, should reset uniforms to default valiues
   reset() {}
 
+  // eslint-disable-next-line complexity
+  _draw(options) {
+    const {
+      logPriority, // Probe log priority, enables Model to do more integrated logging
+
+      framebuffer,
+
+      drawMode = GL.TRIANGLES,
+      vertexCount,
+      offset = 0,
+      isIndexed = false,
+      indexType = GL.UNSIGNED_SHORT,
+      isInstanced = false,
+      instanceCount = 0,
+      start,
+      end
+    } = options;
+
+    if (logPriority !== undefined) {
+      const fb = framebuffer ? framebuffer.id : 'default';
+      const message =
+        `mode=${getKey(this.gl, drawMode)} verts=${vertexCount} ` +
+        `instances=${instanceCount} indexType=${getKey(this.gl, indexType)} ` +
+        `isInstanced=${isInstanced} isIndexed=${isIndexed} ` +
+        `Framebuffer=${fb}`;
+      log.log(logPriority, message)();
+    }
+
+    // Note: async textures set as uniforms might still be loading.
+    // Now that all uniforms have been updated, check if any texture
+    // in the uniforms is not yet initialized, then we don't draw
+    if (!this._areTexturesRenderable()) {
+      return false;
+    }
+
+    if (this._multiDraw(options)) {
+      return true;
+    }
+
+    // Note - Uses polyfilled WebGL2RenderingContext instead of ANGLE extension
+    if (isIndexed && isInstanced) {
+      this.gl.drawElementsInstanced(drawMode, vertexCount, indexType, offset, instanceCount);
+    } else if (isIndexed && isWebGL2(this.gl) && !isNaN(start) && !isNaN(end)) {
+      this.gl.drawRangeElements(drawMode, start, end, vertexCount, indexType, offset);
+    } else if (isIndexed) {
+      this.gl.drawElements(drawMode, vertexCount, indexType, offset);
+    } else if (isInstanced) {
+      this.gl.drawArraysInstanced(drawMode, offset, vertexCount, instanceCount);
+    } else {
+      this.gl.drawArrays(drawMode, offset, vertexCount);
+    }
+
+    return true;
+  }
+
+  // eslint-disable-next-line complexity
+  _multiDraw(options) {
+    const {
+      firsts,
+      firstsOffset = 0,
+      counts,
+      countsOffset = 0,
+      offsets,
+      offsetsOffset = 0,
+      instanceCounts,
+      instanceCountOffset = 0,
+      drawCount = firsts ? firsts.length : 0
+    } = options;
+
+    const isMultidraw =
+      firsts || firstsOffset || counts || countsOffset || offsets || offsetsOffset || drawCount;
+
+    if (!isMultidraw) {
+      return false;
+    }
+
+    const {isIndexed, isInstanced, drawMode, indexType} = options;
+
+    const multiDrawExt = this.gl.getExtension('WEBGL_multi_draw');
+    const multiDrawInstancedExt = this.gl.getExtension('WEBGL_multi_draw_instanced');
+    // TODO - In WebGL2 we could fallback to drawRangeElements for non instanced rendering?
+    assert(multiDrawExt && multiDrawInstancedExt);
+
+    if (isIndexed && isInstanced) {
+      multiDrawInstancedExt.multiDrawElementsInstancedWEBGL(
```

Comment:
Unfortunately it is quite a verbose API, hopefully will minify reasonably so the bundle size impact of supporting it stays small.

Should do before/after measurements...